neuralpredictors/measures/np_functions.py

```diff
@@ -12,10 +19,136 @@ def corr(y1, y2, axis=-1, eps=1e-8, **kwargs):
         y2:      second NumPy array
         axis:    dimension(s) along which the correlation is computed. Any valid NumPy axis spec works here
         eps:     offset to the standard deviation to avoid exploding the correlation due to small division (default 1e-8)
-        **kwargs: passed to final numpy.mean operatoin over standardized y1 * y2
+        **kwargs: passed to final numpy.mean operation over standardized y1 * y2
 
     Returns: correlation array
     """
+
     y1 = (y1 - y1.mean(axis=axis, keepdims=True)) / (y1.std(axis=axis, keepdims=True, ddof=0) + eps)
     y2 = (y2 - y2.mean(axis=axis, keepdims=True)) / (y2.std(axis=axis, keepdims=True, ddof=0) + eps)
     return (y1 * y2).mean(axis=axis, **kwargs)
+
+
+def oracle_corr_corrected(repeated_outputs: ArrayLike) -> np.ndarray:
+    """
+    Compute the corrected oracle correlations per neuron.
+
+    Args:
+        repeated_outputs (array-like): numpy array with shape (images, repeats, responses), or a list containing for each
+            image an array of shape (repeats, responses).
+
+    Returns:
+        array: Corrected oracle correlations per neuron
+    """
+
+    var_noise, var_mean = [], []
+    for repeat in repeated_outputs:
+        var_noise.append(repeat.var(axis=0))
+        var_mean.append(repeat.mean(axis=0))
+    var_noise = np.mean(np.array(var_noise), axis=0)
+    var_mean = np.var(np.array(var_mean), axis=0)
+    return var_mean / np.sqrt(var_mean * (var_mean + var_noise))
+
+
+def oracle_corr(repeated_outputs: ArrayLike) -> np.ndarray:
+    """
+    Compute the oracle correlations per neuron.
+
+    Args:
+        repeated_outputs (array-like): numpy array with shape (images, repeats, responses), or a list containing for each
+            image an array of shape (repeats, responses).
+
+    Returns:
+        array: Oracle correlations per neuron
+    """
+
+    oracles = []
+    for outputs in repeated_outputs:
+        r, n = outputs.shape
+        # compute the mean over repeats, for each neuron
+        mu = outputs.mean(axis=0, keepdims=True)
+        # compute oracle predictor
+        oracle = (mu - outputs / r) * r / (r - 1)
+
+        if np.any(np.isnan(oracle)):
+            logger.warning(
+                "{}% NaNs when calculating the oracle. NaNs will be set to Zero.".format(np.isnan(oracle).mean() * 100)
+            )
+            oracle[np.isnan(oracle)] = 0
+
+        oracles.append(oracle)
+    return corr(np.vstack(repeated_outputs), np.vstack(oracles), axis=0)
+
+
+def explainable_var(repeated_outputs: ArrayLike, eps: int = 1e-9) -> np.ndarray:
```

Comment:
what's the reason for supporting a list of arrays and just simply a full array? Someone could just go ahead and stack the list of arrays into a 3d array right?

Reply:
the lists are of unequal sizes most of the time. that's why they can be lists. the 3d shape is (images, repeats, neurons), and the number of repeats can vary greatly. One could just fill it up with NaNs, or just keep the list structure. Personally I just dont like NaNs so much, because they dont give a sense of the the actual content of the array.

Reply:
Does that comment refer to line 80 or 83? :)

Regarding l. 80 and previous: In the end we vstack anyways. So I am wondering if we couldn't do that in the first place, then avoid the loop (I think this is what @eywalker proposes). Or is there something that I miss @KonstantinWilleke ?

Reply:
👍